### PR TITLE
feat(frontend): add gestion fichiers

### PR DESCRIPTION
**Pull request by UnbearableBear**
All the credits to @revolunet, j'ai siimplement mis un coup de peinture
credits to @lionelB for the ideas and the readme

### CODE REVIEW

File: targets/frontend/src/components/layout/auth.layout.js [hunk 1]
```diff
@@ -8,7 +8,16 @@ import { Header } from "./header";
 import { Nav } from "./Nav";
 import { Stack } from "./Stack";
 
-export function Layout({ children, title }) {
+const Container = ({ noStack, children }) => {
```
Comment (UnbearableBear) [inline targets/frontend/src/components/layout/auth.layout.js:11]: Je suis pas fullstack ! Ca rendait la gestion des "..." buggy

Comment (lionelB) [reply targets/frontend/src/components/layout/auth.layout.js:11]: je suis propose qu'on vire stack (qui servait à gérer l'espacement vertical dans le container et notamment entre les children) et qu'on le rajoute au besoin dans les sous composant du layout.  

Je pense que c'est plus simple si le layout des pages rester le meme

File: targets/frontend/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+Pour chaque environnement, il faudra spécifier, l'url du serveur ainsi que le nom du container.
+
+## Organisation
+
+### cdtn
+
+déploiement | container | instance azure
+- | - | -
+branche | cdtn-dev | dev
+master | cdtn-dev | dev
```
Comment (lionelB) [inline targets/frontend/README.md:13]:
```suggestion
master | cdtn | dev
```
ca me parait plus réaliste que master soi s aligné sur preprod et master de cdtn-admin

File: targets/frontend/src/components/layout/auth.layout.js (hunk 1 above)
Comment (UnbearableBear) [reply targets/frontend/src/components/layout/auth.layout.js:11]: ok je regarde ça

File: .env [hunk 3]
```diff
@@ -12,6 +12,10 @@ SMTP_URL=smtp.url
 SMTP_EMAIL_USER=email
 SMTP_EMAIL_PASSWORD=pass
 
+AZURE_STORAGE_ACCOUNT_KEY=s0m3cr4zyK3y
```
Comment (revolunet) [inline .env:15]:
```suggestion
AZURE_STORAGE_ACCOUNT_KEY=accountKey
```

File: targets/frontend/src/components/layout/auth.layout.js (hunk 1 above)
Comment (UnbearableBear) [reply targets/frontend/src/components/layout/auth.layout.js:11]: https://github.com/SocialGouv/cdtn-admin/pull/238

File: targets/frontend/src/components/layout/Nav.js [hunk 4]
```diff
@@ -115,6 +115,11 @@ export function Nav() {
               Blocs KALI
             </ActiveLink>
           </Li>
+          <Li>
+            <ActiveLink href="/storage/cdtn-test1" passHref>
```
Comment (lionelB) [inline targets/frontend/src/components/layout/Nav.js:119]: le container doit etre `/cdtn` `/cdtn-dev`  
On pourrait coté api forcer le container à cdtn-dev pour les branche / deploiment sur le cluster dev et cdtn pour master ?

File: targets/frontend/src/components/storage/DropZone.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  );
+
+  const { getRootProps, getInputProps, isDragAccept } = useDropzone({
+    accept:
+      "image/jpeg, image/png, application/pdf, application/vnd.openxmlformats-officedocument.wordprocessingml.document",
+    onDrop,
+  });
+
+  if (isDragAccept) {
+    defaultStyles.background = "#d1ffd9"; // some light green
```
Comment (lionelB) [inline targets/frontend/src/components/storage/DropZone.js:43]: I think we can put the color in theme.

File: targets/frontend/src/components/storage/DropZone.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      for (const i in acceptedFiles) {
+        if (acceptedFiles[i] instanceof File) {
+          console.log(acceptedFiles[i].path);
+          console.log(acceptedFiles[i]);
+          formData.append(acceptedFiles[i].path, acceptedFiles[i]);
+        }
+      }
+      onDropCallback(formData).finally(() => {
+        setUploading(false);
+        mutate("files");
```
Comment (lionelB) [inline targets/frontend/src/components/storage/DropZone.js:30]: I propose to call mutate("files") in the drop callback we pass, to let the upper component to handle swr

File: targets/frontend/src/lib/azure.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  StorageSharedKeyCredential,
+} from "@azure/storage-blob";
+import { getToken } from "src/lib/auth/token";
+
+const AZURE_STORAGE_ACCOUNT = {
+  key: process.env.AZURE_STORAGE_ACCOUNT_KEY || "",
+  name: process.env.AZURE_STORAGE_ACCOUNT_NAME || "",
+};
+
+export const uploadFiles = (container, formData) => {
```
Comment (lionelB) [inline targets/frontend/src/lib/azure.js:13]: We could factorize these calls and I wonder if we should move that to the page component and leave in API code that will be use by server api

File: targets/frontend/src/pages/storage/[container]/index.js [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+  const [currentClip, setCurrentClip] = useState("");
+  const [isSearching, setSearching] = useState(false);
+  const [search, setSearch] = useState("");
+  const [sort, setSort] = useState("mostRecent");
+  const [filter, setFilter] = useState("");
+  const [
+    displayedFiles,
+    setDisplayedFiles,
+    setDebouncedDisplayedFiles,
+  ] = useDebouncedState([], 400);
```
Comment (lionelB) [inline targets/frontend/src/pages/storage/[container]/index.js:63]: why use a useDebouncedState ? swr trigger render when it needs to ?

Comment (UnbearableBear) [reply targets/frontend/src/pages/storage/[container]/index.js:63]: It's not about SWR it's about delagging the input when typing. The display rerender is longer than the typing, we don't want the typing to "freeze"

File: targets/frontend/src/components/storage/DropZone.js (hunk 6 above)
Comment (UnbearableBear) [reply targets/frontend/src/components/storage/DropZone.js:30]: Yep it's gonna handle the loading state too. Having a forcer promise callback is too clumsy

File: targets/frontend/src/components/storage/DropZone.js (hunk 5 above)
Comment (UnbearableBear) [reply targets/frontend/src/components/storage/DropZone.js:43]: gonna change for the existing positive green which is fine imho

File: targets/frontend/src/pages/storage/[container]/index.js [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+    const timeout = window.setTimeout(() => {
+      setCurrentClip("");
+    }, 5000);
+    return () => {
+      if (timeout) window.clearTimeout(timeout);
+    };
+  }, [currentClip, setCurrentClip]);
+
+  function copyToClipboard(data) {
+    navigator.clipboard.writeText(data);
```
Comment (lionelB) [inline targets/frontend/src/pages/storage/[container]/index.js:111]: `navigator.clipboard.writeText(data)` return a promise

Could we do 
```
navigator.clipboard.writeText(data).then(() => {
 // code to handle currentClip
}).finally();


```

File: targets/frontend/src/pages/storage/[container]/index.js [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+                      {file.name}
+                    </div>
+                    <div sx={{ fontSize: "small" }}>
+                      <span sx={{ fontWeight: "bold" }}>Poids&nbsp;:</span>{" "}
+                      {prettyBytes(file.contentLength)} | Mise en ligne il y a{" "}
+                      {timeSince(file.lastModified)}
+                    </div>
+                  </div>
+                  {hasClipboardApi && (
+                    <Button
```
Comment (lionelB) [inline targets/frontend/src/pages/storage/[container]/index.js:235]: I think a dedicated ClipBoardButton which handle all the dedicated code (feature detection + localState for clipboard ui status update) would be clearer

File: targets/frontend/src/lib/azure.js (hunk 7 above)
Comment (UnbearableBear) [reply targets/frontend/src/lib/azure.js:13]: techniquement c'est pas trop un component plutot dans lib alors ? Pour la facto bon je pense que ça peut attendre

File: targets/frontend/src/lib/azure.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  const container = getBlobContainer(containerName);
+  const client = container.getBlockBlobClient(stream.name);
+  const blockSize = 4 * 1024 * 1024;
+  const timeout = 30 * 60 * 1000;
+  const concurrency = 20;
+  const options = { abortSignal: AbortController.timeout(timeout) };
+  try {
+    return client.uploadStream(stream, blockSize, concurrency, options);
+  } catch (err) {
+    console.log("Error", err);
```
Comment (lionelB) [inline targets/frontend/src/lib/azure.js:85]: we should throw an error here to forward it to api

File: targets/frontend/src/pages/storage/[container]/index.js (hunk 9 above)
Comment (UnbearableBear) [reply targets/frontend/src/pages/storage/[container]/index.js:111]: the code above is not about async issues it's all about removing the copied state after a while but it's not really necessary at all

File: targets/frontend/src/pages/storage/[container]/index.js (hunk 8 above)
Comment (lionelB) [reply targets/frontend/src/pages/storage/[container]/index.js:63]: as I understand, I would use debouncedState on the filter input and then update the files accordingly

Comment (UnbearableBear) [reply targets/frontend/src/pages/storage/[container]/index.js:63]: Nice ! Way better thanks

File: targets/frontend/src/lib/request.js [hunk 12]
```diff
@@ -8,7 +8,13 @@ export function request(endpoint, { body, ...customConfig } = {}) {
     },
   };
   if (body) {
-    config.body = JSON.stringify(body);
+    if (FormData && body instanceof FormData) {
```
Comment (lionelB) [inline targets/frontend/src/lib/request.js:11]: memo @lionelb:  FormData n'est pas accessible cote serveur du coup ca claque une erreur au refreshToken

**Pull request merged**